zbus/src/address/transport/vsock.rs

```diff
@@ -1,49 +1,67 @@
-use crate::{Error, Result};
-use std::collections::HashMap;
-
-/// A VSOCK D-Bus address.
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub struct Vsock {
-    pub(super) cid: u32,
-    pub(super) port: u32,
+use std::marker::PhantomData;
+
+use super::{percent::decode_percents_str, Address, Error, KeyValFmt, KeyValFmtAdd, Result};
+
+/// `vsock:` D-Bus transport.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Vsock<'a> {
+    // no cid means ANY
+    cid: Option<u32>,
+    // no port means ANY
+    port: Option<u32>,
+    // use a phantom lifetime for eventually future fields and consistency
+    phantom: PhantomData<&'a ()>,
 }
 
-impl Vsock {
-    /// Create a new VSOCK address.
-    pub fn new(cid: u32, port: u32) -> Self {
-        Self { cid, port }
+impl<'a> Vsock<'a> {
+    /// The VSOCK port.
+    pub fn port(&self) -> Option<u32> {
+        self.port
     }
 
-    /// The Client ID.
-    pub fn cid(&self) -> u32 {
+    /// The VSOCK CID.
+    pub fn cid(&self) -> Option<u32> {
         self.cid
     }
+}
 
-    /// The port.
-    pub fn port(&self) -> u32 {
-        self.port
-    }
+impl<'a> TryFrom<&'a Address<'a>> for Vsock<'a> {
```

Comment:
These aren't really conversions. Typically when you implement a conversion from a reference, it's a hint that it's likely not really a conversion. I suggest adding a `Transport` trait in `address::transport` and putting a `from_address` constructor there, as well as moving `key_val_fmt_add` method there (and remove the `KeyValFmtAdd` trait). Also a better name for `key_val_fmt_add` would be nice (or add some docs on it if you can't come up with something better).

Reply:
ok

Reply:
in #1050